File: app/src/main/java/com/example/purepawapp/ui/admin/AdminCategoryAdapter.java
package com.example.purepawapp.ui.admin;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.purepawapp.data.model.Category;
import com.example.purepawapp.databinding.ItemAdminCategoryBinding;

import java.util.function.BiConsumer;
import java.util.function.Consumer;

public class AdminCategoryAdapter extends ListAdapter<CategoryRow, AdminCategoryAdapter.ViewHolder> {

    private final BiConsumer<Category, Boolean> onToggleActive;
    private final Consumer<Category> onEdit;
    private final Consumer<Category> onDelete;

    private static final DiffUtil.ItemCallback<CategoryRow> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull CategoryRow oldItem, @NonNull CategoryRow newItem) {
            return oldItem.getCategory().getId().equals(newItem.getCategory().getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull CategoryRow oldItem, @NonNull CategoryRow newItem) {
            return oldItem.getProductCount() == newItem.getProductCount()
                    && oldItem.getCategory().isActive() == newItem.getCategory().isActive()
                    && oldItem.getCategory().getName().equals(newItem.getCategory().getName());
        }
    };

    public AdminCategoryAdapter(BiConsumer<Category, Boolean> onToggleActive, Consumer<Category> onEdit, Consumer<Category> onDelete) {
        super(DIFF_CALLBACK);
        this.onToggleActive = onToggleActive;
        this.onEdit = onEdit;
        this.onDelete = onDelete;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemAdminCategoryBinding binding = ItemAdminCategoryBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemAdminCategoryBinding binding;

        ViewHolder(ItemAdminCategoryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(CategoryRow row) {
            binding.tvIcon.setText(AdminStatusUi.categoryEmoji(row.getCategory()));
            binding.tvName.setText(row.getCategory().getName());
            binding.tvProductCount.setText(row.getProductCount() + " sản phẩm");
            binding.switchActive.setOnCheckedChangeListener(null);
            binding.switchActive.setChecked(row.getCategory().isActive());
            binding.switchActive.setOnCheckedChangeListener((btn, isChecked) -> onToggleActive.accept(row.getCategory(), isChecked));
            binding.btnEdit.setOnClickListener(v -> onEdit.accept(row.getCategory()));
            binding.btnDelete.setOnClickListener(v -> onDelete.accept(row.getCategory()));
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/admin/AdminCategoryAdapter.kt
package com.example.purepawapp.ui.admin

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.purepawapp.data.model.Category
import com.example.purepawapp.databinding.ItemAdminCategoryBinding

data class CategoryRow(val category: Category, val productCount: Int)

class AdminCategoryAdapter(
    private val onToggleActive: (Category, Boolean) -> Unit,
    private val onEdit: (Category) -> Unit,
    private val onDelete: (Category) -> Unit
) : ListAdapter<CategoryRow, AdminCategoryAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemAdminCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(private val binding: ItemAdminCategoryBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(row: CategoryRow) {
            binding.tvIcon.text = categoryEmoji(row.category)
            binding.tvName.text = row.category.name
            binding.tvProductCount.text = "${row.productCount} sản phẩm"
            binding.switchActive.setOnCheckedChangeListener(null)
            binding.switchActive.isChecked = row.category.isActive
            binding.switchActive.setOnCheckedChangeListener { _, isChecked ->
                onToggleActive(row.category, isChecked)
            }
            binding.btnEdit.setOnClickListener { onEdit(row.category) }
            binding.btnDelete.setOnClickListener { onDelete(row.category) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<CategoryRow>() {
            override fun areItemsTheSame(oldItem: CategoryRow, newItem: CategoryRow): Boolean =
                oldItem.category.id == newItem.category.id
            override fun areContentsTheSame(oldItem: CategoryRow, newItem: CategoryRow): Boolean = oldItem == newItem
        }
    }
}
